### PR TITLE
Custom drawable Symbol Icon

### PR DESCRIPTION
**Pull request by stefankarschti**
Extend the Custom Drawable interface to allow for the addition of symbol icons to the layer. 
The options include:
- texture / image
- texture coordinates
- display size
- anchor point
- rotation around anchor point, in degrees
- scale with map (yes/no)
- pitch with map (yes/no)

ref #1904 

image

### CODE REVIEW

**Review by louwers — CHANGES_REQUESTED** [review]
Can you add a test?

Maybe something that outputs that piece of art?

**Review by mwilsnd — APPROVED** [review]
Looking really good! In a follow up PR, can you add a test that renders an icon across a tile boundary?

**Comment by stefankarschti** [comment]
> Looking really good! In a follow up PR, can you add a test that renders an icon across a tile boundary?

Of course. I'm looking at the current implementation, which anchors the symbol at `(0.5, 0.5)` of tile `(11, 327, 791)` and trying out various scenarios:
`scaleWithMap=true`, `pitchWithMap=false`
image

`scaleWithMap=false`, `pitchWithMap=false`
image

`scaleWithMap=false`, `pitchWithMap=true`
image

`scaleWithMap=true`, `pitchWithMap=true`
image

**Comment by mwilsnd** [comment]
Great, that covers my curiosity.

**Review by TimSylvester — APPROVED** [review]

File: src/mbgl/style/layers/custom_drawable_layer.cpp [hunk 1]
```diff
@@ -181,6 +183,61 @@ class FillDrawableTweaker : public gfx::DrawableTweaker {
...
+        }
+
+        const UnwrappedTileID tileID = drawable.getTileID()->toUnwrapped();
+        mat4 tileMatrix;
+        parameters.state.matrixFor(/*out*/ tileMatrix, tileID);
+
+        const auto matrix = LayerTweaker::getTileMatrix(
+            tileID, parameters, {{0, 0}}, style::TranslateAnchorType::Viewport, false, false, drawable, false);
+
+        static const StringIdentity idDrawableUBOName = stringIndexer().get("CustomSymbolIconDrawableUBO");
```
Comment (alexcristici) [inline src/mbgl/style/layers/custom_drawable_layer.cpp:206]: You will have to update the UBO definition, after https://github.com/maplibre/maplibre-native/pull/1980 will be merged.

File: src/mbgl/style/layers/custom_drawable_layer.cpp [hunk 2]
```diff
@@ -274,15 +336,98 @@ void CustomDrawableLayerHost::Interface::addFill(const GeometryCollection& geome
...
+    builder->setVertexAttributes(std::move(attrs));
+    builder->setRawVertices({}, vertices.elements(), gfx::AttributeDataType::Float2);
+    builder->setSegments(gfx::Triangles(), sharedTriangles, triangleSegments.data(), triangleSegments.size());
+
+    // texture
+    if (symbolOptions.texture) {
+        static const StringIdentity idTextureUniformName = stringIndexer().get("u_texture");
+        if (auto textureLocation = builder->getShader()->getSamplerLocation(idTextureUniformName);
+            textureLocation.has_value()) {
+            builder->setTexture(symbolOptions.texture, *textureLocation);
```
Comment (alexcristici) [inline src/mbgl/style/layers/custom_drawable_layer.cpp:408]: I just merged https://github.com/maplibre/maplibre-native/pull/2107, so you probably need to update this code in order to compile.

**Review by louwers — APPROVED** [review]
Awesome!
